reconcile/aws_saml_roles/integration.py

```diff
@@ -85,40 +177,75 @@ def get_aws_accounts(
+    def get_roles(
         self, query_func: Callable, account_name: str | None = None
-    ) -> list[AWSGroupV1]:
-        """Get all AWS groups with SSO enabled."""
-        data = aws_groups_query(query_func)
-        return [
-            group
-            for group in data.aws_groups or []
-            if integration_is_enabled(self.name, group.account)
-            and (not account_name or group.account.name == account_name)
-            and group.account.sso
-            and group.roles
-            and group.policies
-            and any(role.users for role in group.roles)
-        ]
+    ) -> list[AwsRole]:
+        """Return all roles with AWS account relations."""
+        aws_roles = []
+        for role in roles_query(query_func).roles or []:
+            if not role.aws_groups and not role.user_policies:
+                continue
+
+            user_policies = role.user_policies or []
+            aws_groups = role.aws_groups or []
+            for sso_aws_account in self._filter_aws_accounts(
+                [i.account for i in user_policies + aws_groups],
+                account_name=account_name,
+            ):
+                # AWS limits are checked via pydantic validators
+                custom_policies = []
+                managed_policies = []
+
+                for user_policy in user_policies:
+                    if user_policy.account.uid != sso_aws_account.uid:
+                        continue
+
+                    custom_policies.append(
+                        CustomPolicy(name=user_policy.name, policy=user_policy.policy)
+                    )
+
+                for aws_group in aws_groups:
+                    if aws_group.account.uid != sso_aws_account.uid:
+                        continue
+
+                    managed_policies.extend([
+                        ManagedPolicy(name=p) for p in aws_group.policies or []
+                    ])
+
+                aws_roles.append(
+                    AwsRole(
+                        name=f"{sso_aws_account.uid}-{role.name}",
+                        account=sso_aws_account.name,
+                        custom_policies=custom_policies,
+                        managed_policies=managed_policies,
+                    )
+                )
+
+        return aws_roles
 
     def populate_saml_iam_roles(
-        self, ts: TerrascriptClient, aws_groups: Iterable[AWSGroupV1]
+        self, ts: TerrascriptClient, aws_roles: Iterable[AwsRole]
     ) -> None:
         """Populate the SAML IAM roles."""
-        for group in aws_groups:
-            # aws groups without policies are filtered out by the query
-            # duplicated policies aren't allowed in the same role
-            policies = group.policies or []
-            if len(policies) != len(set(policies)):
-                raise ValueError(
-                    f"Group {group.name} has duplicated policies: {policies}"
+        for role in aws_roles:
+            seen_custom_policies = []
```

Comment:
should use set instead of list for `in` check. btw is duplicate policy already validated before?

Reply:
Fixed in [076aefe](https://github.com/app-sre/qontract-reconcile/pull/4184/commits/076aefeafa9d107cfec95216cc301b023d23f7f5). What do you 

> btw is duplicate policy already validated before?

what do you mean?

Reply:
I saw some validation logic checking duplicate policies, not sure if it's already checked that all policies passed here already deduplicated.

Reply:
The main logic is below. The integration creates a policy once per account, regardless of how often a policy is referenced in other roles.

Reply:
I see, same policy can be refered in multiple roles, but we should only create it once, this is to get uniq policies, this is role level deduplicate, right? The code will reinit `seen_custom_policies` for every role, can 2 different roles for the same account? And it seems to deduplicate by policy name of one role? Is it same as this logic?

```python
        for role in aws_roles:
            unique_policies = {
                custom_policy.name: custom_policy
                for custom_policy in role.custom_policies
            }

            for custom_policy in unique_policies.values():
                ts.populate_iam_policy(
                    account=role.account,
                    name=custom_policy.name,
                    policy=custom_policy.policy,
                )
```

Reply:
Or you mean account level deduplicate

```python
        unique_policies = {
            (role.account, custom_policy.name): custom_policy
            for role in aws_roles
            for custom_policy in role.custom_policies
        }
```

Reply:
Very good catch Di ❤️ I hate [Mock.assert_has_calls](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_has_calls) because it doesn't validate the call count :(
Fixed in the last commit.

Reply:
plus one, I only use `assert_has_calls` when `assert_called_once_with` can't be used, strict check in test is better.